### PR TITLE
Vkj/redesign list.js

### PR DESCRIPTION
**Pull request by vankasteelj**
Changes proposed in this pull request: everything

fix: https://github.com/butterproject/butter-desktop/issues/578

### CODE REVIEW

File: src/app/httpapi.js [hunk 1]
```diff
@@ -454,7 +454,7 @@
             ////////////////////
             server.expose('getgenres', function (args, opt, callback) {
                 switch (App.currentview) {
-                case 'shows':
+                case 'tvshow':
```
Comment (xaiki) [inline src/app/httpapi.js]: you can't case this, it's invalid now.

File: src/app/httpapi.js [hunk 2]
```diff
@@ -479,13 +479,13 @@
...
                 case 'anime':
                     butterCallback(callback, false, {
                         'sorters': App.Config.sorters_tv
                     });
                     break;
-                case 'movies':
+                case 'movie':
```
Comment (xaiki) [inline src/app/httpapi.js]: see my changes to list.js on how to handle this

File: src/app/lib/models/content_item.js [hunk 3]
```diff
@@ -31,7 +31,7 @@
             var torrents = this.get('torrents');
 
             if (!torrents) {
-                console.error('tried to update health, but still no torrents here', this);
+                //console.error('tried to update health, but still no torrents here', this);
```
Comment (xaiki) [inline src/app/lib/models/content_item.js]: this is actually a real issue that should be fixed... if we see this we're trying to render too soon

File: src/app/lib/views/browser/filter_bar.js [hunk 4]
```diff
@@ -163,7 +163,7 @@
 
             if (typeof App.currentview === 'undefined') {
                 App.currentview = activetab;
-                App.previousview = 'movies';
+                App.previousview = 'movie';
```
Comment (xaiki) [inline src/app/lib/views/browser/filter_bar.js]: this should be App.Config.getTabTypes()[0]

File: src/app/lib/views/browser/generic_browser.js [hunk 5]
```diff
@@ -107,8 +107,8 @@
             if (!view) {
                 var activetab;
                 var tabs = {
-                    'TV Series': 'shows',
-                    'Movies': 'movies',
+                    'TV Series': 'tvshow',
+                    'Movies': 'movie',
                     'Anime': 'anime'
```
Comment (xaiki) [inline src/app/lib/views/browser/generic_browser.js]: this should die and use App.Config.getTabs

File: src/app/lib/views/browser/generic_browser.js [hunk 6]
```diff
@@ -118,7 +118,7 @@
                     activetab = AdvSettings.get('startScreen');
                 }
 
-                view = tabs[activetab] || 'movies';
+                view = tabs[activetab] || 'movie';
```
Comment (xaiki) [inline src/app/lib/views/browser/generic_browser.js]: all occurences to 'movie' and 'movies' should die

File: src/app/lib/views/browser/list.js [hunk 7]
```diff
@@ -65,156 +45,139 @@
...
+            Mousetrap.bind('down', this.moveDown.bind(this));
+            Mousetrap.bind('left', this.moveLeft.bind(this));
+            Mousetrap.bind('right', this.moveRight.bind(this));
+            Mousetrap.bind('f', this.toggleSelectedFavourite.bind(this));
+            Mousetrap.bind('w', this.toggleSelectedWatched.bind(this));
+            Mousetrap.bind(['enter', 'space'], this.selectItem.bind(this));
+            Mousetrap.bind(['ctrl+f', 'command+f'], this.focusSearch.bind(this));//FIXME: needs to be moved elsewhere
+            Mousetrap(document.querySelector('input')).bind(['ctrl+f', 'command+f', 'esc'], this.blurSearch.bind(this));//FIXME: needs to be moved elsewhere
+            Mousetrap.bind(['tab', 'shift+tab'], this.switchTab.bind(this));//FIXME: needs to be moved elsewhere
+            Mousetrap.bind(['ctrl+1', 'ctrl+2', 'ctrl+3'], this.switchSpecificTab.bind(this));//FIXME: needs to be moved elsewhere
```
Comment (xaiki) [inline src/app/lib/views/browser/list.js]: need to add as many ctrl+i than we have tabs in App.Config.getTabTypes()

File: src/app/lib/views/browser/list.js [hunk 8]
```diff
@@ -229,87 +192,98 @@
...
+                    '</span>' +
+                    '<div class="loading-container">' +
+                        '<div class="sk-folding-cube">' +
+                              '<div class="sk-cube1 sk-cube"></div>' +
+                              '<div class="sk-cube2 sk-cube"></div>' +
+                              '<div class="sk-cube4 sk-cube"></div>' +
+                              '<div class="sk-cube3 sk-cube"></div>' +
+                        '</div>' +
+                    '</div>' +
+                '</div>';
```
Comment (xaiki) [inline src/app/lib/views/browser/list.js]: load this from a template, too much HTML for js...

File: src/app/lib/views/browser/list.js [hunk 9]
```diff
@@ -229,87 +192,98 @@
...
-                    if (this.collection.hasMore && !this.collection.filter.keywords && this.collection.state !== 'error' && this.collection.length !== 0) {
-                        $('#load-more-item').css('display', 'inline-block').click(function () {
-                            $('#load-more-item').off('click');
-                            self.collection.fetchMore();
-                        });
-                    }
-                }
-                break;
+            if (
+                ['movie', 'tvshow', 'anime'].indexOf(App.currentview) !== -1
```
Comment (xaiki) [inline src/app/lib/views/browser/list.js]: you can't do that. need to check App.Config.getTabTypes()

File: src/app/lib/views/browser/list.js [hunk 10]
```diff
@@ -318,133 +292,116 @@
...
             var currentPosition = this.$el.scrollTop() + this.$el.height();
 
-            if (this.collection.state === 'loaded' &&
-                (currentPosition / totalHeight) > SCROLL_MORE) {
-                this.collection.fetchMore();
-            }
+            return (
+                this.collection.state === 'loaded' 
+                && (currentPosition / totalHeight) > SCROLL_MORE
+            ) ? this.collection.fetchMore() : true;
```
Comment (xaiki) [inline src/app/lib/views/browser/list.js]: split that, it's unclear in that return statement

File: src/app/lib/views/main_window.js [hunk 11]
```diff
@@ -249,6 +249,7 @@
                 .indexOf(newTab) !== -1) {
                 // XXX hack until we get something better
                 this['show' + newTab.capitalize()]();
+                App.currentview = newTab;
```
Comment (xaiki) [inline src/app/lib/views/main_window.js]: move hat earlier so that it's not repeated

File: src/app/templates/browser/list.tpl [hunk 12]
```diff
@@ -1,9 +1,9 @@
...
-		  <div class="sk-cube1 sk-cube"></div>
-		  <div class="sk-cube2 sk-cube"></div>
-		  <div class="sk-cube4 sk-cube"></div>
-		  <div class="sk-cube3 sk-cube"></div>
-		</div>
+        <div class="sk-cube1 sk-cube"></div>
+        <div class="sk-cube2 sk-cube"></div>
+        <div class="sk-cube4 sk-cube"></div>
+        <div class="sk-cube3 sk-cube"></div>
+    </div>
```
Comment (xaiki) [inline src/app/templates/browser/list.tpl]: split that to another template.

File: src/app/httpapi.js (hunk 1 above)
Comment (vankasteelj) [inline src/app/httpapi.js]: then the entire httpapi needs to be rewritten. it won't be in this PR. see #576

File: src/app/httpapi.js (hunk 2 above)
Comment (vankasteelj) [inline src/app/httpapi.js]: not related to this PR, i'm just fixing what needs fixing, I'm not rewriting something else than list. see #576

File: src/app/lib/views/browser/filter_bar.js (hunk 4 above)
Comment (vankasteelj) [inline src/app/lib/views/browser/filter_bar.js]: done

File: src/app/lib/views/browser/generic_browser.js (hunk 5 above)
Comment (vankasteelj) [inline src/app/lib/views/browser/generic_browser.js]: not related to this PR, i'm just making it work atm -> #576

File: src/app/lib/views/browser/generic_browser.js (hunk 6 above)
Comment (vankasteelj) [inline src/app/lib/views/browser/generic_browser.js]: not related to this PR, i'm just making it work atm -> #576

File: src/app/lib/models/content_item.js (hunk 3 above)
Comment (vankasteelj) [inline src/app/lib/models/content_item.js]: it's a structural issue then

File: src/app/lib/views/browser/list.js (hunk 10 above)
Comment (vankasteelj) [inline src/app/lib/views/browser/list.js]: done

File: src/app/lib/views/browser/list.js (hunk 9 above)
Comment (vankasteelj) [inline src/app/lib/views/browser/list.js]: done

File: src/app/templates/browser/list.tpl (hunk 12 above)
Comment (vankasteelj) [inline src/app/templates/browser/list.tpl]: I disagree, it's the loadmore for list, it has to be in list, no?

File: src/app/lib/views/main_window.js (hunk 11 above)
Comment (vankasteelj) [inline src/app/lib/views/main_window.js]: done

File: src/app/lib/views/browser/list.js (hunk 7 above)
Comment (vankasteelj) [inline src/app/lib/views/browser/list.js]: done

File: src/app/lib/views/browser/list.js (hunk 8 above)
Comment (vankasteelj) [inline src/app/lib/views/browser/list.js]: you said yourself you didn't wanted to bang the dom. If we let backbone handle this, it will.

Comment (xaiki) [inline src/app/lib/views/browser/list.js]: we can put it in a template and do the loading ourselves. really too much js, and it's already in list.tpl so we need to keep both in sync.

File: src/app/templates/browser/list.tpl (hunk 12 above)
Comment (xaiki) [inline src/app/templates/browser/list.tpl]: but we *then* have it in the view.

**Comment by xaiki** [comment]
merging this so i can fix 'movies' and stuff.

File: src/app/lib/views/browser/list.js (hunk 8 above)
Comment (vankasteelj) [inline src/app/lib/views/browser/list.js]: i don't care as long as I don't have to fix it a second time.

File: src/app/lib/providers/watchlist.js [hunk 13]
```diff
@@ -183,26 +183,7 @@
...
-                title: data.episode_title,
-                tvdb_id: data.episode_id,
-                first_aired: new Date(data.episode_aired) / 1000,
-                date_based: false,
-                torrents: {},
-            }];
-        }
-
-        return Promise.resolve(formatted);
+        return Promise.reject('No details for watchlist');
```
Comment (vankasteelj) [inline src/app/lib/providers/watchlist.js]: I think it lacks a few mandatory things (torrents, type), the key names arent the same (genres vs genre, etc), plus it doesnt work for series as there is only the "next episode" in watchlist, not the entire show+episodes list.

Comment (xaiki) [inline src/app/lib/providers/watchlist.js]: we need to cache this... it's a TODO for butter-providers

**Pull request merged**